x/ccv/provider/keeper/throttle.go

```diff
@@ -19,54 +19,56 @@ import (
 func (k Keeper) HandleThrottleQueues(ctx sdktypes.Context) {
 
 	meter := k.GetSlashMeter(ctx)
-
-	// Don't start iterating if meter is negative in value
+	// Return if meter is negative in value
 	if meter.IsNegative() {
 		return
 	}
 
-	handledGlobalEntries := []providertypes.GlobalSlashEntry{}
-
-	// Iterate through ordered (by received time) global slash entries from any consumer chain
-	k.IterateGlobalSlashEntries(ctx, func(entry providertypes.GlobalSlashEntry) (stop bool) {
-
-		// Obtain validator from the provider's consensus address.
-		// Note: if validator is not found or unbonded, this will be handled appropriately in HandleSlashPacket
-		val, found := k.stakingKeeper.GetValidatorByConsAddr(ctx, entry.ProviderValConsAddr)
-
-		// Obtain the validator power relevant to the slash packet that's about to be handled
-		// (this power will be removed via jailing or tombstoning)
-		var valPower int64
-		if !found || val.IsJailed() {
-			// If validator is not found, or found but jailed, it's power is 0. This path is explicitly defined since the
-			// staking keeper's LastValidatorPower values are not updated till the staking keeper's endblocker.
-			valPower = 0
-		} else {
-			valPower = k.stakingKeeper.GetLastValidatorPower(ctx, val.GetOperator())
-		}
+	// Obtain all global slash entries, where only some of them may be handled in this method,
+	// depending on the value of the slash meter.
+	allEntries := k.GetAllGlobalSlashEntries(ctx)
+	handledEntries := []providertypes.GlobalSlashEntry{}
 
-		// Subtract this power from the slash meter
-		meter = meter.Sub(sdktypes.NewInt(valPower))
+	for _, globalEntry := range allEntries {
+		// Subtract voting power that will be jailed/tombstoned from the slash meter
+		meter = meter.Sub(k.GetEffectiveValPower(ctx, globalEntry.ProviderValConsAddr))
 
-		// Handle slash and any trailing vsc matured packet data instances by passing in
+		// Handle one slash and any trailing vsc matured packet data instances by passing in
 		// chainID and appropriate callbacks, relevant packet data is deleted in this method.
-		k.HandlePacketDataForChain(ctx, entry.ConsumerChainID, k.HandleSlashPacket, k.HandleVSCMaturedPacket)
+		k.HandlePacketDataForChain(ctx, globalEntry.ConsumerChainID, k.HandleSlashPacket, k.HandleVSCMaturedPacket)
+		handledEntries = append(handledEntries, globalEntry)
 
-		// Store handled global entry to be deleted after iteration is completed
-		handledGlobalEntries = append(handledGlobalEntries, entry)
-
-		// Do not handle anymore global slash entries if the meter is negative in value
-		stop = meter.IsNegative()
-		return stop
-	})
+		// don't handle any more global entries if meter becomes negative in value
+		if meter.IsNegative() {
+			break
```

Comment:
Include this break into `GetAllGlobalSlashEntries` as meter is not modified in `HandlePacketDataForChain`. This also means that `handledEntries` is no longer needed.